src/decode/ykpt/packet_parser/mod.rs

```diff
@@ -0,0 +1,303 @@
+            self.state
+        )
+        .unwrap();
+        write!(&mut s, "{}", self.byte_stream_str(8, ", ")).unwrap();
+        HWTracerError::TraceParseError(s)
+    }
+
+    /// Attempt to parse a packet of the specified `PacketKind`.
+    fn parse_kind(&mut self, kind: PacketKind) -> Option<Packet> {
+        let bits = BitSlice::from_slice(self.bytes).ok()?;
+        let parse_res = match kind {
+            PacketKind::PSB => {
+                read_to_packet!(PSBPacket, bits, Packet::PSB)
+            }
+            PacketKind::CBR => read_to_packet!(CBRPacket, bits, Packet::CBR),
+            PacketKind::PSBEND => read_to_packet!(PSBENDPacket, bits, Packet::PSBEND),
+            PacketKind::PAD => read_to_packet!(PADPacket, bits, Packet::PAD),
+            PacketKind::MODE => read_to_packet!(MODEPacket, bits, Packet::MODE),
+            PacketKind::TIPPGE => {
+                read_to_packet_tip!(TIPPGEPacket, bits, Packet::TIPPGE, self.prev_tip)
+            }
+            PacketKind::TIPPGD => {
+                read_to_packet_tip!(TIPPGDPacket, bits, Packet::TIPPGD, self.prev_tip)
+            }
+            PacketKind::ShortTNT => read_to_packet!(ShortTNTPacket, bits, Packet::ShortTNT),
+            PacketKind::LongTNT => read_to_packet!(LongTNTPacket, bits, Packet::LongTNT),
+            PacketKind::TIP => read_to_packet_tip!(TIPPacket, bits, Packet::TIP, self.prev_tip),
+            PacketKind::FUP => read_to_packet_tip!(FUPPacket, bits, Packet::FUP, self.prev_tip),
+            PacketKind::CYC => read_to_packet!(CYCPacket, bits, Packet::CYC),
+        };
+        if let Ok((remain, pkt)) = parse_res {
+            self.bytes = remain.as_raw_slice();
+            Some(pkt)
+        } else {
+            None
+        }
+    }
+
+    /// Attempt to parse a packet for the current parser state.
+    fn parse_state(&mut self) -> Result<Packet, HWTracerError> {
+        for kind in self.state.valid_packets() {
+            if let Some(pkt) = self.parse_kind(*kind) {
+                if *kind == PacketKind::PSBEND {
+                    self.state = PacketParserState::Normal;
+                }
+                return Ok(pkt);
+            }
+        }
+        Err(self.mk_parse_error())
+    }
+
+    /// Returns a string showing a binary formatted peek at the next `nbytes` bytes of
+    /// `self.bytes`. Bytes in the output are separated by `sep`.
+    ///
+    /// This is used to format error messages, but is also useful when debugging.
+    fn byte_stream_str(&self, nbytes: usize, sep: &str) -> String {
+        use std::cmp::min;
+        let nbytes = min(nbytes, self.bytes.len());
+        let mut vals = Vec::new();
+        for i in 0..nbytes {
+            vals.push(format!("{:08b}", self.bytes[i]));
+        }
+
+        if self.bytes.len() > nbytes {
+            vals.push("...".to_owned());
+        }
+
+        let mut ret = String::new();
+        write!(ret, "{}", vals.join(sep)).unwrap();
```

Comment:
The `write!` is pointless: `vals.join` creates a string which we then `write!` into an empty string!

Reply:
Oops. That's a left-over from an earlier design where the output string was passed in. Wlil fix.

Reply:
Fixed in [3567abf](https://github.com/ykjit/hwtracer/pull/105/commits/3567abfb722b6106c105a07bc9a0b4ceda47d3a4)